idp/facebook.go

```diff
@@ -0,0 +1,195 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package idp
+
+import (
+    "bytes"
+    "encoding/json"
+    "fmt"
+    "io"
+    "net/http"
+    "net/url"
+    "time"
+
+    "golang.org/x/oauth2"
+)
+
+type FacebookIdProvider struct {
+    Client *http.Client
+    Config *oauth2.Config
+}
+
+func NewFacebookIdProvider(clientId string, clientSecret string, redirectUrl string) *FacebookIdProvider {
+    idp := &FacebookIdProvider{}
+
+    config := idp.getConfig(clientId, clientSecret, redirectUrl)
+    idp.Config = config
+
+    return idp
+}
+
+func (idp *FacebookIdProvider) SetHttpClient(client *http.Client) {
+    idp.Client = client
+}
+
+// getConfig return a point of Config, which describes a typical 3-legged OAuth2 flow
+func (idp *FacebookIdProvider) getConfig(clientId string, clientSecret string, redirectUrl string) *oauth2.Config {
+    var endpoint = oauth2.Endpoint{
+        TokenURL: "https://graph.facebook.com/oauth/access_token",
+    }
+
+    var config = &oauth2.Config{
+        Scopes:       []string{"email,public_profile"},
+        Endpoint:     endpoint,
+        ClientID:     clientId,
+        ClientSecret: clientSecret,
+        RedirectURL:  redirectUrl,
+    }
+
+    return config
+}
+
+type FacebookAccessToken struct {
+    AccessToken string `json:"access_token"` //Interface call credentials
+    TokenType   string `json:"token_type"`   //Access token type
+    ExpiresIn   int64  `json:"expires_in"`   //access_token interface call credential timeout time, unit (seconds)
+}
+
+type FacebookCheckToken struct {
+    Data string `json:"data"`
+}
+
+// Get more detail via: https://developers.facebook.com/docs/facebook-login/manually-build-a-login-flow#checktoken
+type FacebookCheckTokenData struct {
+    UserId string `json:"user_id"`
+}
+
+// GetToken use code get access_token (*operation of getting code ought to be done in front)
+// get more detail via: https://developers.facebook.com/docs/facebook-login/manually-build-a-login-flow#confirm
+func (idp *FacebookIdProvider) GetToken(code string) (*oauth2.Token, error) {
+    params := url.Values{}
+    params.Add("client_id", idp.Config.ClientID)
+    params.Add("redirect_uri", idp.Config.RedirectURL)
+    params.Add("client_secret", idp.Config.ClientSecret)
+    params.Add("code", code)
+
+    accessTokenUrl := fmt.Sprintf("https://graph.facebook.com/oauth/access_token?%s", params.Encode())
+
+    accessTokenResp, err := idp.GetUrlResp(accessTokenUrl)
+    if err != nil {
+        return nil, err
+    }
+
+    var facebookAccessToken FacebookAccessToken
+    if err = json.Unmarshal([]byte(accessTokenResp), &facebookAccessToken); err != nil {
+        return nil, err
+    }
+
+    token := oauth2.Token{
+        AccessToken: facebookAccessToken.AccessToken,
+        TokenType:   "FacebookAccessToken",
+        Expiry:      time.Time{},
+    }
+
+    return &token, nil
+}
+
+//{
```

Comment:
Is this real response?

Reply:
The format of the real response looks like this, and the value of `id` `name` `url` and `email` value is modified here